#!/bin/bash

## saga
module load Nextflow/24.04.2

nf-core pipelines download sarek --revision 3.5.1 \
    --outdir ./sarek_offline \
    --container-system singularity \
    --compress none




# 2. Download annotation caches


### need old python version
conda create -n gsutil_env python=3.10 -y
conda activate gsutil_env
pip install gsutil

mkdir -p gatk_hg38

# Core reference files
gsutil cp gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta ./gatk_hg38
gsutil cp gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.fai ./gatk_hg38
gsutil cp gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dict ./gatk_hg38

# BWA indices (will be auto-generated by Sarek if not present)
gsutil cp gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.fasta.64.* ./gatk_hg38

# Known sites VCF files
gsutil cp gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf.gz ./gatk_hg38
gsutil cp gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf.gz.tbi ./gatk_hg38
gsutil cp gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.known_indels.vcf.gz ./gatk_hg38
gsutil cp gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.known_indels.vcf.gz.tbi ./gatk_hg38
gsutil cp gs://gcp-public-data--broad-references/hg38/v0/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz ./gatk_hg38
gsutil cp gs://gcp-public-data--broad-references/hg38/v0/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz.tbi ./gatk_hg38

# Somatic variants
gsutil cp gs://gatk-best-practices/somatic-hg38/af-only-gnomad.hg38.vcf.gz ./gatk_hg38
gsutil cp gs://gatk-best-practices/somatic-hg38/af-only-gnomad.hg38.vcf.gz.tbi ./gatk_hg38

